### ISSUE TITLE
False positive: NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE on annotated parameterized type

### ISSUE THREAD

**Issue by wrprice** [spotbugs/spotbugs#927]
I'm getting what I consider to be a false-positive NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE on the following example code:

```
package com.example.spotbugs;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import javax.annotation.meta.TypeQualifierNickname;

/**
 * False positive NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE due to type parameter annotation.
 * The report says:
 * <pre>
 * container must be non-null but is marked as nullable
 *
 * In class com.example.spotbugs.NpParamNonNullFalsePositive
 * In method com.example.spotbugs.NpParamNonNullFalsePositive.contrivedToString(NpParamNonNullFalsePositive$Container)
 * Parameter container
 * At NpParamNonNullFalsePositive.java:[lines 27-28]
 * </pre>
 */
public class NpParamNonNullFalsePositive {

    public <T> String contrivedToString(Container<@Nullable T> container) {
        // Argument `container` is unconditionally dereferenced, but *it* is NOT marked @Nullable.
        // The @Nullable annotation is on its parameterized type <T>, i.e. getValue() could return null.
        Object val = container.getValue();
        return val == null ? "null" : val.toString(); // getValue() result is explicitly checked
    }

    public static interface Container <T> {
        T getValue();
    }

    @Retention(RetentionPolicy.RUNTIME)
    @Target({ElementType.PARAMETER, ElementType.TYPE_USE})  // NOTE: target includes TYPE_USE
    @Documented
    @TypeQualifierNickname      // NOTE: both annotations (on this line...
    @javax.annotation.Nullable  //       ... and this line) are required to trigger the condition
    public static @interface Nullable {
    }
}
```

May sound similar to #844 but the difference here is the (custom) `@Nullable` annotation being placed within the _parameterized type_ of the method parameter, _not the method parameter itself_.  In other words, this is specific to the annotation's use as enabled by [`@Target`](https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.6.4.1) including [`ElementType.TYPE_USE`](https://docs.oracle.com/javase/8/docs/api/java/lang/annotation/ElementType.html#TYPE_USE)

In my real production code, I have such a custom annotation defined similar to the above explicitly because the standard `@javax.annotation.Nullable` annotation _does not_ support `TYPE_USE`, which was added in Java 8.

I'm using SpotBugs 3.1.12, with version 1.7.1 of the Gradle plugin, in the following environment:

```
------------------------------------------------------------
Gradle 5.2.1
------------------------------------------------------------

Build time:   2019-02-08 19:00:10 UTC
Revision:     f02764e074c32ee8851a4e1877dd1fea8ffb7183

Kotlin DSL:   1.1.3
Kotlin:       1.3.20
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.13 compiled on July 10 2018
JVM:          11.0.3 (Azul Systems, Inc. 11.0.3+7-LTS)
OS:           Windows 10 10.0 amd64
```

**Comment by olegshtch** [comment]
I get it same for `@org.​checkerframework.​checker.​nullness.​qual.Nullable` which supports `TYPE_USE` with 4.0.6 spotbugs.

**Comment by olegshtch** [comment]
```java
package org.test.testgradle1;

import java.util.HashMap;
import java.util.Map;
import org.checkerframework.checker.nullness.qual.Nullable;

public class Main {

	public static void main(String[] args) {
		Map<String, @Nullable String> params = parseParams("a=1;b=2");
		doSomething(params);
	}
	
	private static void doSomething(Map<String, @Nullable String> params) {
		for (Map.Entry<String, @Nullable String> e : params.entrySet()) {
			System.out.print(e);
		}
	}
	
	protected static Map<String, @Nullable String> parseParams(String parameters) {
		Map<String, @Nullable String> result = new HashMap<>();
		String[] parts = parameters.split(";");
		for (String part : parts) {
			String[] line = part.split("=", 2);
			String key = line[0].trim();
			String value = null;
			if (line.length > 1) {
				value = line[1].trim();
			}
			result.put(key, value);
		}
		return result;
	}

}
```

Gives error:
```
Bug type NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE (click for details) 
In class org.test.testgradle1.Main
In method org.test.testgradle1.Main.doSomething(Map)
Parameter params
At Main.java:[lines 15-18]
```

**Comment by jpschewe** [comment]
The same problem exists with marking the values of an array, but not the array itself.

```
  private static boolean isHeaderLine(final @Nullable String[] line) {
    boolean retval = false;
    for (final String element : line) {
      if (TEAM_NUMBER_HEADER.equals(element)) {
        retval = true;
      }
    }

    return retval;
  }
```

**Comment by gtoison** [comment]
This should be fixed by #2502
